{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os \n",
    "import Bio\n",
    "from collections import Counter\n",
    "import re\n",
    "from itertools import chain\n",
    "from functools import reduce\n",
    "import pandas as pd\n",
    "\n",
    "def kmer(seq:Bio.SeqRecord.SeqRecord, k:int=20, stride:int=2):\n",
    "    \"\"\"\n",
    "    用于切分kmer序列，\n",
    "\n",
    "    Parameters:\n",
    "      seq: SeqRecord类型或者str\n",
    "      k - kmer大小\n",
    "      stride - 步长\n",
    "\n",
    "    Returns:\n",
    "        list,[kmer_1, kmer_2,..., kmer_n]\n",
    "\n",
    "    Raises:\n",
    "        KeyError - raises an exception\n",
    "    \"\"\"\n",
    "    \n",
    "    if isinstance(seq, Bio.SeqRecord.SeqRecord):\n",
    "        seq = seq.seq\n",
    "    elif isinstance(seq, str):\n",
    "        seq = seq\n",
    "    else:\n",
    "        print(\"传入Seqrecord或者str类型\")\n",
    "        raise TypeError\n",
    "        \n",
    "    length = len(seq)\n",
    "    \n",
    "    return [str(seq[i: i + k]) for i in range(0, length, stride) if i + k <= length ]\n",
    "        \n",
    "        \n",
    "def readFastaFileFromGivenPath(path):\n",
    "    \"\"\"\n",
    "    Return:\n",
    "        一个包含所有fasta序列的迭代器\n",
    "    \"\"\"\n",
    "    pathList = [os.path.join(path, file) for file in os.listdir(path) if re.match(f\".+fasta\", file) ]\n",
    "    fastaIterList = [Bio.SeqIO.parse(fastaPath, \"fasta\") for fastaPath in pathList]\n",
    "    return chain(*fastaIterList)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {},
   "outputs": [],
   "source": [
    "def kmer(seq:Bio.SeqRecord.SeqRecord, k:int=20, stride:int=2):\n",
    "    \"\"\"\n",
    "    用于切分kmer序列，\n",
    "\n",
    "    Parameters:\n",
    "      seq: SeqRecord类型或者str\n",
    "      k - kmer大小\n",
    "      stride - 步长\n",
    "\n",
    "    Returns:\n",
    "        list,[kmer_1, kmer_2,..., kmer_n]\n",
    "\n",
    "    Raises:\n",
    "        KeyError - raises an exception\n",
    "    \"\"\"\n",
    "    \n",
    "    if isinstance(seq, Bio.SeqRecord.SeqRecord):\n",
    "        seq = seq.seq\n",
    "    elif isinstance(seq, str):\n",
    "        seq = seq\n",
    "    else:\n",
    "        print(\"传入Seqrecord或者str类型\")\n",
    "        raise TypeError\n",
    "        \n",
    "    length = len(seq)\n",
    "    \n",
    "    return [str(seq[i: i + k]) for i in range(0, length, stride) if i + k <= length ]\n",
    "        \n",
    "        \n",
    "def readFastaFileFromGivenPath(path):\n",
    "    \"\"\"\n",
    "    Return:\n",
    "        一个包含所有fasta序列的迭代器\n",
    "    \"\"\"\n",
    "    pathList = [os.path.join(path, file) for file in os.listdir(path) if re.match(f\".+fasta\", file) ]\n",
    "    fastaIterList = [Bio.SeqIO.parse(fastaPath, \"fasta\") for fastaPath in pathList]\n",
    "    return chain(*fastaIterList)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 120,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "def genomes2kmer(path, output = \"./htt.csv\", k = 20, stride = 10):\n",
    "    \"\"\"\n",
    "    用去切分基因组序列成kmer片段，按照stride进行切割。\n",
    "    切分：[genome[i:i+k] for i in range(i, genomeLength, stride)]\n",
    "    最后返回的是一个csv文件，第一列是kmer序列，剩余列是各个基因组中该kmer的频次\n",
    "    \n",
    "    Parameters:\n",
    "      path - 指定基因组序列存储的位置，该目录下可以有多个.fasta文件\n",
    "      output - 输出目录\n",
    "      k - kmer大小\n",
    "      stride - 步长\n",
    "\n",
    "    Returns:\n",
    "        csv file\n",
    "\n",
    "    \"\"\"\n",
    "    CounterList = []\n",
    "    genomeIdList = []\n",
    "    for fasta in readFastaFileFromGivenPath(path):\n",
    "        genomeIdList.append(fasta.id)\n",
    "        kmerSeq = kmer(fasta, k, stride)\n",
    "        CounterList.append(Counter(kmerSeq))\n",
    "    \n",
    "    allKmerCounter = dict(reduce(lambda x, y : x + y, CounterList))\n",
    "    \n",
    "    result = []\n",
    "    for key in allKmerCounter.keys():\n",
    "        distribution = [genomedic.get(key, 0) for genomedic in CounterList]\n",
    "\n",
    "        result.append(pd.DataFrame([[key, *distribution]], columns=[f\"{k}-mer\", *genomeIdList]))   \n",
    "    \n",
    "    result = pd.concat(result)\n",
    "    result.to_csv(output,index=0)\n",
    "    return result\n",
    "        \n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 121,
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>20-mer</th>\n",
       "      <th>NZ_CP007136.1</th>\n",
       "      <th>NZ_CP007133.1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AGCTTTTCATTCTGACTGCA</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TCTGACTGCAACGGGCAATA</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ACGGGCAATATGTCTCTGTG</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TGTCTCTGTGTGGATTAAAA</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TGGATTAAAAAAAGAGTCTC</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AAAGAGTCTCTGACAGCAGC</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TGACAGCAGCTTCTGAACTG</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TTCTGAACTGGTTACCTGCC</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GTTACCTGCCGTGAGTAAAT</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GTGAGTAAATTAAAATTTTA</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TAAAATTTTATTGACTTAGG</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TTGACTTAGGTCACTAAATA</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TCACTAAATACTTTAACCAA</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CTTTAACCAATATAGGCATA</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TATAGGCATAGCGCACAGAC</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GCGCACAGACAGATAAAAAT</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AGATAAAAATTACAGAGTAC</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TACAGAGTACACAACATCCA</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ACAACATCCATGAAACGCAT</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TGAAACGCATTAGCACCACC</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TAGCACCACCATTACCACCA</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ATTACCACCACCATCACCAT</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CCATCACCATTACCACAGGT</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TACCACAGGTAACGGTGCGG</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AACGGTGCGGGCTGACGCGT</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GCTGACGCGTACAGGAAACA</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ACAGGAAACACAGAAAAAAG</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CAGAAAAAAGCCCGCACCTG</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CCCGCACCTGACAGTGCGGG</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ACAGTGCGGGCTTTTTTTTC</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CTTTTTTTTCGACCAAAGGT</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GACCAAAGGTAACGAGGTAA</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AACGAGGTAACAACCATGCG</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CAACCATGCGAGTGTTGAAG</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AGTGTTGAAGTTCGGCGGTA</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TTCGGCGGTACATCAGTAGC</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CATCAGTAGCAAATGCAGAA</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AAATGCAGAACGTTTTCTGC</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CGTTTTCTGCGGGTTGCCGA</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 20-mer  NZ_CP007136.1  NZ_CP007133.1\n",
       "0  AGCTTTTCATTCTGACTGCA              1              1\n",
       "0  TCTGACTGCAACGGGCAATA              1              1\n",
       "0  ACGGGCAATATGTCTCTGTG              1              1\n",
       "0  TGTCTCTGTGTGGATTAAAA              1              1\n",
       "0  TGGATTAAAAAAAGAGTCTC              1              1\n",
       "0  AAAGAGTCTCTGACAGCAGC              1              1\n",
       "0  TGACAGCAGCTTCTGAACTG              1              1\n",
       "0  TTCTGAACTGGTTACCTGCC              1              1\n",
       "0  GTTACCTGCCGTGAGTAAAT              1              1\n",
       "0  GTGAGTAAATTAAAATTTTA              1              1\n",
       "0  TAAAATTTTATTGACTTAGG              1              1\n",
       "0  TTGACTTAGGTCACTAAATA              1              1\n",
       "0  TCACTAAATACTTTAACCAA              1              1\n",
       "0  CTTTAACCAATATAGGCATA              1              1\n",
       "0  TATAGGCATAGCGCACAGAC              1              1\n",
       "0  GCGCACAGACAGATAAAAAT              1              1\n",
       "0  AGATAAAAATTACAGAGTAC              1              1\n",
       "0  TACAGAGTACACAACATCCA              1              1\n",
       "0  ACAACATCCATGAAACGCAT              1              1\n",
       "0  TGAAACGCATTAGCACCACC              1              1\n",
       "0  TAGCACCACCATTACCACCA              1              1\n",
       "0  ATTACCACCACCATCACCAT              1              1\n",
       "0  CCATCACCATTACCACAGGT              1              1\n",
       "0  TACCACAGGTAACGGTGCGG              1              1\n",
       "0  AACGGTGCGGGCTGACGCGT              1              1\n",
       "0  GCTGACGCGTACAGGAAACA              1              1\n",
       "0  ACAGGAAACACAGAAAAAAG              1              1\n",
       "0  CAGAAAAAAGCCCGCACCTG              1              1\n",
       "0  CCCGCACCTGACAGTGCGGG              1              1\n",
       "0  ACAGTGCGGGCTTTTTTTTC              1              1\n",
       "0  CTTTTTTTTCGACCAAAGGT              1              1\n",
       "0  GACCAAAGGTAACGAGGTAA              0              1\n",
       "0  AACGAGGTAACAACCATGCG              0              1\n",
       "0  CAACCATGCGAGTGTTGAAG              0              1\n",
       "0  AGTGTTGAAGTTCGGCGGTA              0              1\n",
       "0  TTCGGCGGTACATCAGTAGC              0              1\n",
       "0  CATCAGTAGCAAATGCAGAA              0              1\n",
       "0  AAATGCAGAACGTTTTCTGC              0              1\n",
       "0  CGTTTTCTGCGGGTTGCCGA              0              1"
      ]
     },
     "execution_count": 121,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "a = htt(\"./\")\n",
    "a"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "RNASeq",
   "language": "python",
   "name": "rnaseq"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
